java-squid/src/main/java/org/sonar/java/cfg/LiveVariables.java

```diff
@@ -0,0 +1,176 @@
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Set;
+
+public class
+    LiveVariables {
+
+  private final CFG cfg;
+  private final Map<CFG.Block, Set<Symbol>> out = new HashMap<>();
+
+  private LiveVariables(CFG cfg) {
+    this.cfg = cfg;
+  }
+
+  public Set<Symbol> getOut(CFG.Block block) {
+    return out.get(block);
+  }
+
+  public static LiveVariables analyze(CFG cfg) {
+    LiveVariables liveVariables = new LiveVariables(cfg);
+    final Map<CFG.Block, Set<Symbol>> in = new HashMap<>();
+    // Generate kill/gen for each block in isolation
+    Map<CFG.Block, Set<Symbol>> kill = new HashMap<>();
+    Map<CFG.Block, Set<Symbol>> gen = new HashMap<>();
+    for (CFG.Block block : liveVariables.cfg.reversedBlocks()) {
+      Set<Symbol> blockKill = new HashSet<>();
+      Set<Symbol> blockGen = new HashSet<>();
+      liveVariables.processBlockElements(block, blockKill, blockGen);
+      kill.put(block, blockKill);
+      gen.put(block, blockGen);
+    }
+    liveVariables.analyzeCFG(in, kill, gen);
+    // out of exit block are empty by definition.
+    if (!liveVariables.out.get(liveVariables.cfg.reversedBlocks().get(0)).isEmpty()) {
+      throw new IllegalStateException("Out of exit block should be empty");
+    }
+
+    // Make things immutable.
+    for (Map.Entry<CFG.Block, Set<Symbol>> blockSetEntry : liveVariables.out.entrySet()) {
+      blockSetEntry.setValue(ImmutableSet.copyOf(blockSetEntry.getValue()));
+    }
+
+    return liveVariables;
+  }
+
+  private void analyzeCFG(Map<CFG.Block, Set<Symbol>> in, Map<CFG.Block, Set<Symbol>> kill, Map<CFG.Block, Set<Symbol>> gen) {
+    Deque<CFG.Block> workList = new LinkedList<>();
+    workList.addAll(cfg.reversedBlocks());
+    while (!workList.isEmpty()) {
+      CFG.Block block = workList.removeFirst();
+
+      Set<Symbol> blockOut = out.get(block);
+      if (blockOut == null) {
+        blockOut = new HashSet<>();
+        out.put(block, blockOut);
+      }
+      for (CFG.Block successor : block.successors()) {
+        Set<Symbol> inOfSuccessor = in.get(successor);
+        if (inOfSuccessor != null) {
+          blockOut.addAll(inOfSuccessor);
+        }
+      }
+      // in = gen and (out - kill)
+      Set<Symbol> newIn = new HashSet<>(gen.get(block));
+      newIn.addAll(Sets.difference(blockOut, kill.get(block)));
+
+      if (newIn.equals(in.get(block))) {
+        continue;
+      }
+      in.put(block, newIn);
+      for (CFG.Block predecessor : block.predecessors()) {
+        workList.addLast(predecessor);
+      }
+    }
+  }
+
+  private void processBlockElements(CFG.Block block, Set<Symbol> blockKill, Set<Symbol> blockGen) {
+    // process elements from bottom to top
+    Set<Tree> assignmentLHS = new HashSet<>();
+    for (Tree element : Lists.reverse(block.elements())) {
+      Symbol symbol;
+      switch (element.kind()) {
+        case ASSIGNMENT:
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 11 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1151)